### PR TITLE
Feature/#66 base site url

### PR DESCRIPTION
**Pull request by royrico**
Hi, I've worked to implement the ability to load private-bower on a "siteBaseURL", enabling the site to be moved to a sub directory.   This helps with hosting the site on a shared internal server or hosting behind a reverse proxy.

For the configuration file, I moved port and hostName options, under a group, called server.  This is where added the config for siteBaseURL.    This groups everything logically together.   I also ensured that, if people still had the port & hostName on the root, it would pick it up there, so that it would not break anyone's existing configuration.

if the siteBaseURL is defined, private-bower will add one additional route, which is to redirect / to the specified URL.    So that people can still type in localhost:port when testing.

Also, fixed some inconsistencies with the 'default' heroku registry urls in the default config, and in the configurationManager.    Both are now the same, and use secure protocol.

Thanks!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -171,7 +176,7 @@ public ones will show up just fine.
 > http://localhost:5678/
 
 ##Project
-Create .bowerrc file with content:
+Within your project, you can create a .bowerrc file containing:
```
Comment (Hacklone) [inline README.md]: I think creating the .bowerrc file is mandatory, so the word ```can``` can be misleading.

File: lib/application.js [hunk 2]
```diff
@@ -39,18 +41,25 @@ module.exports = function Application() {
     function _serveStatic(staticPath) {
         _addMiddleware(_staticHandler(staticPath));
     }
-    
+
     function _addMiddleware(middleware) {
-        _serverApp.use(middleware);
+        if (_siteBaseUrl) {
+            _serverApp.use(_siteBaseUrl,middleware);
```
Comment (Hacklone) [inline lib/application.js]: missing space :)

File: lib/api/infrastructure/controller.js [hunk 3]
```diff
@@ -39,7 +39,10 @@ Controller.prototype = {
             this.router.delete(path, middleware);
         }
     },
-    bind: function(app) {
+    bind: function(app,siteBasePath) {
```
Comment (Hacklone) [inline lib/api/infrastructure/controller.js]: missing space

File: lib/application.js [hunk 4]
```diff
@@ -39,18 +41,25 @@ module.exports = function Application() {
...
+        } else {
```
Comment (Hacklone) [inline lib/application.js]: missing EOL

File: lib/application.js [hunk 5]
```diff
@@ -39,18 +41,25 @@ module.exports = function Application() {
...
+            _serverApp.use(middleware);
+        }
     }
-    
+
+    function _addMiddlewareWithMount(route,middleware) {
```
Comment (Hacklone) [inline lib/application.js]: spaces

File: lib/application.js [hunk 6]
```diff
@@ -39,18 +41,25 @@ module.exports = function Application() {
...
+    }
+
     function _loadControllers(controllersRoot) {
         fs.readdirSync(controllersRoot).forEach(loadControllerAtByName);
         
         function loadControllerAtByName(controllerPath) {
             var controller = require(path.join(controllersRoot, controllerPath));
-            
-            controller.bind(_serverApp);
+            controller.bind(_serverApp,_siteBaseUrl);
```
Comment (Hacklone) [inline lib/application.js]: space

File: lib/main.js [hunk 7]
```diff
@@ -63,14 +63,20 @@ module.exports = function Main() {
 
         application.loadControllers(path.join(utils.dirname, 'api/controllers'));
 
-        application.addMiddleware(function(err, req, res, next) {
+        if ( _config.server.siteBaseURL ) {
```
Comment (Hacklone) [inline lib/main.js]: too much spaces :)

**Comment by Hacklone** [comment]
This is great, could you run the tests if they are running fine?
```gulp test```

Thanks, this is a great help :)

File: README.md (hunk 1 above)
Comment (royrico) [inline README.md]: updated, will check in everything in one shot.

**Comment by royrico** [comment]
I ran the tests last night and they worked just fine. I ran them today, after my quick fixes, and one of them is failing. I looked into it, however, and it's failing because the angular library was updated on github, but hasn't reflected on rawgit just yet.

git:feature/#66-baseSiteUrl ✔ ◒  $ gulp test
[17:08:11] Warning: gulp version mismatch:
[17:08:11] Global gulp is 3.9.0
[17:08:11] Local gulp is 3.8.11
[17:08:12] Using gulpfile ~/Code/amex-enterprise-bower/node_modules/private-bower/gulpfile.js
[17:08:12] Starting 'test'...
[17:08:12] Starting 'unit-test'...
[bower]  27/10/15 17:08:12   private-bower server started
[bower]  27/10/15 17:08:12   private-bower server started
[bower]  27/10/15 17:08:12   private-bower server started
[bower]  27/10/15 17:08:12   private-bower server started
[bower]  27/10/15 17:08:12   private-bower server started
[bower]  27/10/15 17:08:12   private-bower server started
[bower]  27/10/15 17:08:12   private-bower server started
[bower]  27/10/15 17:08:12   private-bower server started
[bower]  27/10/15 17:08:13   private-bower server started
[bower]  27/10/15 17:08:13   private-bower server started
[bower]  27/10/15 17:08:13   private-bower server started
[bower]  27/10/15 17:08:13   private-bower server started
[bower]  27/10/15 17:08:13   private-bower server started
[bower]  27/10/15 17:08:13   private-bower server started
[bower]  27/10/15 17:08:13   private-bower server started
 33  -_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-__,------,
 0   -_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-__|  /\_/\
 2   -_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_~|_( o .o)
     -_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_ ""  ""

  33 passing (908ms)
  2 pending

[17:08:13] Finished 'unit-test' after 1.3 s
[17:08:13] Starting 'feature-test'...
[bower]  27/10/15 17:08:13   Refreshing public packages...
[bower]  27/10/15 17:08:13   private-bower server started
[bower]  27/10/15 17:08:13   Registered package: angular
[bower]  27/10/15 17:08:15   Shutting down private-bower
 1   -__|  /\_/\
 0   -_~|_( x .x)
     -_ ""  ""


  0 passing (2s)
  1 failing

  1) PackageDetails Get package details should get the bower.json of the package:

      AssertionError: expected '{"name":"angular","version":"1.5.0-build.4337+sha.a995ee1","main":"./angular.js","ignore":[],"dependencies":{}}' to equal '{"name":"angular","version":"1.5.0-build.4334+sha.ffb6b2f","main":"./angular.js","ignore":[],"dependencies":{}}'
      + expected - actual

      -{"name":"angular","version":"1.5.0-build.4337+sha.a995ee1","main":"./angular.js","ignore":[],"dependencies":{}}
      +{"name":"angular","version":"1.5.0-build.4334+sha.ffb6b2f","main":"./angular.js","ignore":[],"dependencies":{}}

      at features/packageDetails.spec.js:41:56
      at Object._catch [as catch] (features/assets/utils.js:7:13)
      at features/packageDetails.spec.js:40:28
      at bound (domain.js:254:14)
      at runBound (domain.js:267:12)
      at tryCatcher (node_modules/bluebird/js/main/util.js:26:23)
      at Promise._settlePromiseFromHandler (node_modules/bluebird/js/main/promise.js:507:31)
      at Promise._settlePromiseAt (node_modules/bluebird/js/main/promise.js:581:18)
      at Async._drainQueue (node_modules/bluebird/js/main/async.js:128:12)
      at Async._drainQueues (node_modules/bluebird/js/main/async.js:133:10)
      at Immediate.Async.drainQueues [as _onImmediate] (node_modules/bluebird/js/main/async.js:15:14)



[17:08:15] 'feature-test' errored after 1.73 s
[17:08:15] Error in plugin 'gulp-mocha'
Message:
    1 test failed.
[17:08:15] 'test' errored after 3.04 s
[17:08:15] Error in plugin 'run-sequence'
Message:
    An error occured in task 'feature-test'.
[bower]  27/10/15 17:08:17   Loaded public packages
[bower]  27/10/15 17:08:17   Public registry cached


### i changed it to work with font awesome, and the tests passed:

rrico1@orion:~/Code/private-bower
git:feature/#66-baseSiteUrl ✗ ⚑ ◒  $ gulp test
[20:18:35] Warning: gulp version mismatch:
[20:18:35] Global gulp is 3.9.0
[20:18:35] Local gulp is 3.8.11
[20:18:36] Using gulpfile ~/Code/private-bower/gulpfile.js
[20:18:36] Starting 'test'...
[20:18:36] Starting 'unit-test'...
[bower]  27/10/15 20:18:36   private-bower server started
[bower]  27/10/15 20:18:36   private-bower server started
[bower]  27/10/15 20:18:37   private-bower server started
[bower]  27/10/15 20:18:37   private-bower server started
[bower]  27/10/15 20:18:37   private-bower server started
[bower]  27/10/15 20:18:37   private-bower server started
[bower]  27/10/15 20:18:37   private-bower server started
[bower]  27/10/15 20:18:37   private-bower server started
[bower]  27/10/15 20:18:37   private-bower server started
[bower]  27/10/15 20:18:37   private-bower server started
[bower]  27/10/15 20:18:37   private-bower server started
[bower]  27/10/15 20:18:37   private-bower server started
[bower]  27/10/15 20:18:37   private-bower server started
[bower]  27/10/15 20:18:37   private-bower server started
[bower]  27/10/15 20:18:37   private-bower server started
 33  -_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-__,------,
 0   -_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-__|  /\_/\
 2   -_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_~|_( o .o)
     -_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_ ""  ""

  33 passing (883ms)
  2 pending

[20:18:37] Finished 'unit-test' after 1.25 s
[20:18:37] Starting 'feature-test'...
[bower]  27/10/15 20:18:37   Refreshing public packages...
[bower]  27/10/15 20:18:38   Cached public packages loaded
[bower]  27/10/15 20:18:38   private-bower server started
[bower]  27/10/15 20:18:38   Registered package: fontawesome
[bower]  27/10/15 20:18:40   Shutting down private-bower
 0   -__|  /\_/\
 0   -_~|_( ^ .^)
     -_ ""  ""


  1 passing (3s)

[20:18:40] Finished 'feature-test' after 2.87 s
[20:18:40] Finished 'test' after 4.12 s
[bower]  27/10/15 20:18:41   Loaded public packages
[bower]  27/10/15 20:18:42   Public registry cached

**Pull request merged**